File: Java/InfixToPostfix.java
import java.util.Stack; 
import java.util.Scanner;
  
public class InfixToPostfix 
{ 
    // A utility function to return precedence of a given operator 
    // Higher returned value means higher precedence 
    static int Prec(char ch) 
    { 
        switch (ch) 
        { 
        case '+': 
        case '-': 
            return 1; 
       
        case '*': 
        case '/': 
            return 2; 
       
        case '^': 
            return 3; 
            
        default:
            return;
        } 
        return -1; 
    } 
       
    // The main method that converts given infix expression 
    // to postfix expression.  
    static String infixToPostfix(String exp) 
    { 
        // initializing empty String for result 
        String result = new String(""); 
          
        // initializing empty stack 
        Stack<Character> stack = new Stack<>(); 
          
        for (int i = 0; i<exp.length(); ++i) 
        { 
            char c = exp.charAt(i); 
              
             // If the scanned character is an operand, add it to output. 
            if (Character.isLetterOrDigit(c)) 
                result += c; 
               
            // If the scanned character is an '(', push it to the stack. 
            else if (c == '(') 
                stack.push(c); 
              
            //  If the scanned character is an ')', pop and output from the stack  
            // until an '(' is encountered. 
            else if (c == ')') 
            { 
                while (!stack.isEmpty() && stack.peek() != '(') 
                    result += stack.pop(); 
                  
                if (!stack.isEmpty() && stack.peek() != '(') 
                    return "Invalid Expression"; // invalid expression                 
                else
                    stack.pop(); 
            } 
            // when an operator is encountered
            else  
            { 
                while (!stack.isEmpty() && Prec(c) <= Prec(stack.peek())){ 
                    if(stack.peek() == '(') 
                        return "Invalid Expression"; 
                    result += stack.pop(); 
             } 
                stack.push(c); 
            } 
       
        } 
       
        // pop all the operators from the stack 
        while (!stack.isEmpty()){ 
            if(stack.peek() == '(') 
                return "Invalid Expression"; 
            result += stack.pop(); 
         } 
        return result; 
    } 
    
    public static void main(String[] args)  
    { 
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the infix string to be converted to postfix:");
        String exp = sc.nextLine(); 
        System.out.println(infixToPostfix(exp)); 
    } 
} 

File: C/infix_to_postfix.c
/* Infix to Postfix Conversion */

#include <stdio.h>
#include <string.h>
#include <math.h>
void post(char inf[], char postfix[]);
int pop();
void push(int n);
int arr[20];
int top = -1, var[256];
int main(void)
{
    char inf[20], postfix[20], c;
    int val;
    var['^'] = 5;
    var['*'] = 3;
    var['/'] = 3;
    var['+'] = 2;
    var['-'] = 2;
    var['('] = 1;
    printf("Enter the Infix expression\n");
    scanf("%s", inf);
    post(inf, postfix);
    printf("Postfix Expression: %s\n", postfix);
}
int pop()
{
    top--;
    return arr[top + 1];
}
void push(int n)
{
    top++;
    arr[top] = n;
}
void post(char inf[], char postfix[])
{
    int i = 0, j = 0, k = strlen(inf);
    push('(');
    inf[k] = ')';
    inf[k + 1] = '\0';
    while (inf[i] != '\0')
    {
        if (inf[i] == '(')
            push('(');
        else if (inf[i] >= 'a' && inf[i] <= 'z')
        {
            postfix[j] = inf[i];
            j++;
        }
        else if (inf[i] == '/' || inf[i] == '-' || inf[i] == '+' || inf[i] == '*' || inf[i] == '^')
        {
            if (arr[top] == '(')
                push(inf[i]);
            else
            {
                if (var[inf[i]] <= var[arr[top]])
                {
                    while (var[arr[top]] >= var[inf[i]])
                    {
                        postfix[j] = pop();
                        j++;
                    }
                    push(inf[i]);
                }
                else
                    push(inf[i]);
            }
        }
        else if (inf[i] == ')')
        {
            while (arr[top] != '(')
            {
                postfix[j] = pop();
                j++;
            }
            top--;
        }
        if (inf[i + 1] == '\0')
            postfix[j] = '\0';
        i++;
    }
}

/* Test Cases

a+b*(c^d-e)^(f+g*h)-i

a*(b+c)/d

Output:

abcd^e-fgh*+^*+i-

abc+*d/

*/
